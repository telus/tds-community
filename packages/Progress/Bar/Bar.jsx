import React from 'react'
import PropTypes from 'prop-types'

import { safeRest } from '@tds/util-helpers'

import styles from '../Progress.scss'

const MIN_NON_ZERO_PROGRESS_PERCENTAGE = 5
const MAX_PROGRESS_PERCENTAGE = 100

const constrainNonZeroPercentage = (minNonZeroPercentage, maxPercentage, percentage) => {
  if (percentage <= 0) return 0
  else if (percentage < minNonZeroPercentage) return minNonZeroPercentage
  else if (percentage > maxPercentage) return maxPercentage
  return percentage
}

const Bar = ({ percentage, variant, ...rest }) => {
  const percent = constrainNonZeroPercentage(
    MIN_NON_ZERO_PROGRESS_PERCENTAGE,
    MAX_PROGRESS_PERCENTAGE,
    percentage
  )
  const zIndex = 100 - percent // z-index order is inverse of percentage
  return (
    <div
      {...safeRest(rest)}
      className={[styles.progressBar, styles[variant]].join(' ')}
      style={{ width: `${percent}%`, zIndex }}
    />
  )
}

Bar.propTypes = {
  /**
   * Specifies how much of the task has been completed
   */
  percentage: PropTypes.number.isRequired,
  variant: PropTypes.oneOf(['primary', 'secondary', 'error', 'disabled']),
}

Bar.defaultProps = {
  variant: 'primary',
}

export default Bar
