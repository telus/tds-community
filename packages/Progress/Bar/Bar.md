Bar present amount of completion for a task as a linear bar. The **percentage** property
accepts a number representing percentage of completion.

By default, `Progress.Bar` will display the positive variant. Use the `variant` prop to change the style between `positive`, `negative`, and `disabled`.

```jsx
<Progress>
  <Progress.Bar percentage={40} a11yLabel="Data Used" />
</Progress>
```

```jsx
<Progress>
  <Progress.Bar percentage={40} variant="negative" a11yLabel="Data Used" />
</Progress>
```
