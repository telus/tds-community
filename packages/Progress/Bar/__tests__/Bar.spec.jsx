import React from 'react'
import { shallow } from 'enzyme'

import Bar from '../Bar'

describe('Progress', () => {
  const doShallow = (props = { percentage: 5 }) => shallow(<Bar {...props} />)

  it('renders', () => {
    const progressBar = doShallow()

    expect(progressBar).toMatchSnapshot()
  })

  it('does other things', () => {
    const progressBar = doShallow()

    expect(progressBar).toExist()
  })

  it('passes additional attributes to the element', () => {
    const progressBar = doShallow({ id: 'the-id', 'data-some-attr': 'some value' })

    expect(progressBar).toHaveProp('id', 'the-id')
    expect(progressBar).toHaveProp('data-some-attr', 'some value')
  })

  it('does not allow custom CSS', () => {
    const progressBar = doShallow({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(progressBar).not.toHaveProp('className', 'my-custom-class')
    expect(progressBar).not.toHaveProp('style', { color: 'hotpink' })
  })

  it('has a width that is a percentage proportional to its value', () => {
    const progressBar = doShallow({ percentage: 50 }).find('.progressBar')
    expect(progressBar).toHaveStyle({ width: '50%' })
  })

  it('has a width of 5% if value is equal or less than 5', () => {
    const progressBar = doShallow({ percentage: 5 }).find('.progressBar')
    expect(progressBar).toHaveStyle({ width: '5%' })
  })
})
