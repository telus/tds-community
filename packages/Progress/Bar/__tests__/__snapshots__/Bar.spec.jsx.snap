// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress renders 1`] = `
<div
  className="progressBar primary"
  style={
    Object {
      "width": "5%",
      "zIndex": 95,
    }
  }
/>
`;
